src/mongo/db/audit/audit.cpp

```diff
@@ -0,0 +1,901 @@
+
+#include "audit_options.h"
+#include "audit_file.h"
+
+#define PERCONA_AUDIT_STUB {}
+
+namespace mongo {
+
+namespace audit {
+
+    NOINLINE_DECL void realexit( ExitCode rc ) {
+#ifdef _COVERAGE
+        // Need to make sure coverage data is properly flushed before exit.
+        // It appears that ::_exit() does not do this.
+        log() << "calling regular ::exit() so coverage data may flush..." << std::endl;
+        ::exit( rc );
+#else
+        ::_exit( rc );
+#endif
+    }
+
+    // Writable interface for audit events
+    class WritableAuditLog : public logger::AuditLog {
+    public:
+        virtual ~WritableAuditLog() {}
+        virtual void append(const BSONObj &obj) = 0;
+        virtual void rotate() = 0;
+    };
+
+    // Writes audit events to a json file
+    class JSONAuditLog : public WritableAuditLog {
+        bool ioErrorShouldRetry(int errcode) {
+            return (errcode == EAGAIN ||
+                    errcode == EWOULDBLOCK ||
+                    errcode == EINTR);
+        }
+
+    public:
+        JSONAuditLog(const std::string &file, const BSONObj &filter)
+            : _file(new AuditFile), 
+              _matcher(filter.getOwned()), 
+              _fileName(file),
+              _mutex("auditFileMutex") {
+            _file->open(file.c_str(), false, false);
+        }
+
+        virtual void append(const BSONObj &obj) {
+            if (_matcher.matches(obj)) {
+                const std::string str = mongoutils::str::stream() << obj.jsonString() << "\n";
+
+                // mongo::File does not have an "atomic append" operation.
+                // As such, with a rwlock we are vulnerable to a race
+                // where we get the length of the file, then try to pwrite
+                // at that offset.  If another write beats us to pwrite,
+                // we'll overwrite that audit data when our write goes
+                // through.
+                //
+                // Somewhere, we need a mutex around grabbing the file
+                // offset and trying to write to it (even if this were in
+                // the kernel, the synchronization is still there).  This
+                // is a good enough place as any.
+                //
+                // We don't need the mutex around fsync, except to protect against concurrent
+                // logRotate destroying our pointer.  Welp.
+                SimpleMutex::scoped_lock lck(_mutex);
+
+                // If pwrite performs a partial write, we don't want to
+                // muck about figuring out how much it did write (hard to
+                // get out of the File abstraction) and then carefully
+                // writing the rest.  Easier to calculate the position
+                // first, then repeatedly write to that position if we
+                // have to retry.
+                fileofs pos = _file->len();
+
+                int writeRet;
+                for (int retries = 10; retries > 0; --retries) {
```

Comment:
This part of code was ported from tokumx-audit-enterprise repository (https://github.com/percona/tokumx-audit-enterprise). There it was authored on Sep 29, 2014 as MXE-20 task. I don't know if MXE-20 can be viewed somewhere. History is here: https://github.com/percona/tokumx-audit-enterprise/commits/master/audit.cpp
In my understanding this code's purpose is to resolve issues when logRotate() switches current log file simultaneously with this write.